Implement tools to produce network graphs in MemoryNet examples

philipr-za opened the pull request:
## Description
This PR adds some utility functions to take snapshots of the network as a graph at any point during a memory net run. The utility functions provide the ability to take snapshots in separate named sequences if you want to graph different networks or elements of the networks separately. The graph files are stored in the temporary directory. Finally a Python script is included that will read the graphs for each named sequence and render them using a Force-Directed graph layout algorithm to the provided (or default) output folder.

Supported:
 - Network graphs with all connections and/or neighbour connections
- Message propagation onto of a network graph

## How Has This Been Tested?
This is not production code and is part of the Dht Examples to be used for debugging. 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

delta1 reviewed (approved):
Running tests locally

delta1 reviewed (changes requested):
Build is failing? running `cargo test`

```
error[E0308]: mismatched types
   --> comms/dht/examples/memorynet_graph_network_join_single_seed.rs:108:22
    |
108 |                 vec![seed_node[0].to_peer()],
    |                      ^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::sync::Arc`, found struct `tari_comms::peer_manager::peer::Peer`
    |
    = note: expected struct `std::sync::Arc<tari_comms::peer_manager::node_identity::NodeIdentity>`
               found struct `tari_comms::peer_manager::peer::Peer`

error[E0599]: no variant or associated item named `NetworkGraph` found for enum `graphing_utilities::utilities::PythonRenderType` in the current scope
   --> comms/dht/examples/memorynet_graph_network_join_single_seed.rs:161:101
    |
161 |         run_python_network_graph_render("after_base_node_join", graph_output_dir, PythonRenderType::NetworkGraph)
    |                                                                                                     ^^^^^^^^^^^^
    |                                                                                                     |
    |                                                                                                     variant or associated item not found in `graphing_utilities::utilities::PythonRenderType`
    |                                                                                                     help: there is a variant with a similar name: `NetworkGraphFull`
    | 
   ::: comms/dht/examples/graphing_utilities/utilities.rs:271:1
    |
271 | pub enum PythonRenderType {
    | ------------------------- variant or associated item `NetworkGraph` not found here

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `tari_comms_dht`.
```

philipr-za commented:
doh, was supposed to axe that guy

sdbondi commented on comms/src/peer_manager/node_id.rs:
```diff
@@ -104,6 +105,19 @@ impl TryFrom<&[u8]> for XorDistance {
+    type Error = String;
+
+    fn try_from(value: XorDistance) -> Result<Self, Self::Error> {
+        if value.0.len() > 16 {
+            return Err("XorDistance has too many bytes to be converted to U128".to_string());
+        }
+        let mut bytes: [u8; 16] = [0u8; 16];
+        bytes.iter_mut().set_from(value.0.iter().cloned());
+        Ok(u128::from_be_bytes(bytes))
```
If you use this technique, you can remove the `itertools` dependency

```suggestion
         if XorDistance::byte_length() > 16 {
            return Err("XorDistance has too many bytes to be converted to U128".to_string());
        }
        let slice = value.as_bytes();
        let mut bytes: [u8; 16] = [0u8; 16];
        bytes[..XorDistance::byte_length()].copy_from_slice(&slice[..XorDistance::byte_length()]);
        Ok(u128::from_be_bytes(bytes))
```

```rust
    /// Returns the number of bytes required to represent the `XorDistance`
    pub const fn byte_length() -> usize {
        NODE_XOR_DISTANCE_ARRAY_SIZE
    }
```

philipr-za replied:
But I love itertools! alright done

sdbondi commented on comms/Cargo.toml:
```diff
@@ -39,6 +39,7 @@ tokio = {version="~0.2.19", features=["blocking", "time", "tcp", "dns", "sync",
 tokio-util = {version="0.2.0", features=["codec"]}
 tower= "0.3.1"
 yamux = "=0.4.7"
+itertools = "0.9.0"
```
You can remove this dependency (see comment in comms/src/peer_manager/node_id.rs)

sdbondi reviewed (commented):
Great stuff - having a visual representation of the DHT workings is super helpful.

Just one small comment about removing `itertools`

sdbondi reviewed (dismissed):
Awesome stuff 🥇

stringhandler commented:
Just some compile errors in release

philipr-za committed:
Implement tools to produce network graphs in memory net examples

This PR adds some utility functions to take snapshots of the network as a graph at any point during a memory net run. The utility functions provide the ability to take snapshots in separate named sequences if you want to graph different networks or elements of the networks separately. The graph files are stored in the temporary directory. Finally a Python script is included that will read the graphs for each named sequence and render them using a Force-Directed graph layout algorithm to the provided (or default) output folder.

Supported:
 - Network graphs with all connections and/or neighbour connections
- Message propagation onto of a network graph

philipr-za commented:
Fixed
